### PR TITLE
Load Datadog.Trace.ClrProfiler.Managed via the AppDomain.AssemblyResolve event

### PR DESCRIPTION
**Pull request by zacharycmontoya**
Product changes:
- In `Datadog.Trace.ClrProfiler.Managed.Loader` add the `AppDomain.AssemblyResolve` event handler to load managed profiler dependencies from disk where the profiler is installed. This would remove the need to install files into the GAC or add the `Datadog.Trace.ClrProfiler.Managed` NuGet package to receive auto-instrumentation.
- Stop adding assembly references to the managed profiler code in `ModuleLoadFinished` and instead do that operation in `JITCompilationStarted`. Because of this delayed reference, the runtime will not be able to resolve the assembly reference automatically (perhaps avoiding the SecurityGrant exceptions we have seen in the past!) and will rely on the newly implemented `AppDomain.AssemblyResolve` event to find the managed profiler code.
- Build `Datadog.Trace.ClrProfiler.Managed.Loader` for `netcoreapp2.0` and `net45` because reading environment variables (to obtain profiler file path) is only supported on `netstandard1.3` and higher.

Build changes:
- Combine common settings in `Directory.Build.props` for reproductions, reproduction-dependencies, and samples into `./Test.Common.props`
- For Windows-ready xUnit integration tests, add the `RunOnWindows=True` trait
- Produce managed profiler assets by running `dotnet publish Datadog.Trace.ClrProfiler.Managed` for each target framework
- Stop building `Datadog.Trace.ClrProfiler.Managed.Loader.csproj` inside `Datadog.Trace.proj`'s `BuildCpp` target and instead build it separately whenever we build `Datadog.Trace.ClrProfiler.Native`

(edit: formatting)

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs [hunk 1]
```diff
@@ -14,20 +15,48 @@ public class Startup
...
+        private static Assembly ResolveManagedProfiler(object sender, ResolveEventArgs args)
+        {
+            string tracerDirectory = Environment.GetEnvironmentVariable("DD_DOTNET_TRACER_HOME");
+
+            if (!string.IsNullOrEmpty(tracerDirectory))
+            {
+#if NETSTANDARD2_0
+                string tracerFrameworkDirectory = "netstandard2.0";
+#else
+                string tracerFrameworkDirectory = "net45";
```
Comment (colin-higgins) [inline src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs]: We may want a directory for 461+ because of the Sigil-vNext conditional references. Thoughts?

Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs]: .NET Framework 4.6.1 implements .NET Standard 2.0, so we can load the version targeting `netstandard2.0` in an application running on `net461` or above. The version targeting `net45` is only for apps running on .NET Framework 4.5 - 4.6.

The only reason we target `net461` separately in our NuGet packages is because [it's recommended](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting#multi-targeting) to support devs using older tooling:

> CONSIDER adding a target for `net461` when you're offering a `netstandard2.0` target.
>
> Using .NET Standard 2.0 from .NET Framework has some issues that were addressed in .NET Framework 4.7.2. You can improve the experience for developers that are still on .NET Framework 4.6.1 - 4.7.1 by offering them a binary that is built for .NET Framework 4.6.1.

Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs]: We still need to add the run-time checks to load the correct assembly, though.

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs]: After Lucas's comments, is the sentiment that we should detect the .NET Framework version and use `netstandard2.0` if we're on .NET Framework 4.6.1+? If so, do you guys have a preference on which of these mechanisms to use to identify the framework version? Just checking the CLR version doesn't give us enough info, which is why all these other solutions exist: https://stackoverflow.com/questions/951856/is-there-an-easy-way-to-check-the-net-framework-version

Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs]: Sorry, after some research, I realized that what I suggested is probably a bad idea. Even though `net461` supports `netstandard2.0`, it does so by deploying additional libraries with the app, because `net461` does _not_ implement all of the APIs in `netstandard2.0`. See [this article](https://devblogs.microsoft.com/dotnet/net-framework-4-7-1-runtime-and-compiler-features/ ) for more details.

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs]: Ok, I've updated the logic here to use either the `net45` assets or the `net461` assets for .NET Framework apps. Let me know what your thoughts are on the implementation.

**Comment by kgorro** [comment]
@zacharycmontoya this issue happens both in master and this branch. The scenario is when you create a self hosted app(32bit) and you run web app in that app in 64bit. Write access violation exist for symbolsSize.

**Comment by zacharycmontoya** [comment]
I see, thanks for bringing this to our attention! Could you please create a new issue so we can properly track that and provide some detailed repro steps so we can easily reproduce it?

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 2]
```diff
@@ -173,16 +173,16 @@ HRESULT STDMETHODCALLTYPE CorProfiler::ModuleLoadFinished(ModuleID module_id,
...
-            module_info.assembly.app_domain_name);
+          " AppDomain ", module_info.assembly.app_domain_id, " ",
+          module_info.assembly.app_domain_name);
   }
 
   AppDomainID app_domain_id = module_info.assembly.app_domain_id;
 
   // Identify the AppDomain ID of mscorlib which will be the Shared Domain
   // because mscorlib is always a domain-neutral assembly
   if (!corlib_module_loaded && (module_info.assembly.name == "mscorlib"_W ||
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: I'm worried that this shared domain logic will cause us to lose traces once merged so I think we'll need to modify one of our samples to be a "MultiDomainHost" and count Ado.NET traces. But this should be a followup PR because this PR is already large enough.

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            var tracerHomeDirectory = Environment.GetEnvironmentVariable("DD_DOTNET_TRACER_HOME") ?? string.Empty;
+            return Path.Combine(tracerHomeDirectory, tracerFrameworkDirectory);
+        }
+
+        private static Assembly AssemblyResolve_ManagedProfilerDependencies(object sender, ResolveEventArgs args)
+        {
+            string assemblyName = new AssemblyName(args.Name).Name;
+            var path = Path.Combine(ManagedProfilerDirectory, $"{assemblyName}.dll");
+
+            if (assemblyName.ToLower().Contains("datadog.trace") && File.Exists(path))
```
Comment (colin-higgins) [inline src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs]: Would it be better to use our public key(s)?

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 4]
```diff
@@ -462,11 +443,47 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
...
+             caller.name, "()");
+        continue;
+      }
+
+      const MetadataBuilder metadata_builder(*module_metadata, module,
+                                             module_metadata->metadata_import, module_metadata->metadata_emit,
+                                             module_metadata->assembly_import, module_metadata->assembly_emit);
+
+      // for each wrapper assembly, emit an assembly reference
+      hr = metadata_builder.EmitAssemblyRef(
```
Comment (colin-higgins) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Do we even need to emit before knowing we are going to replace anything?

Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Would it be reasonable to store a flag on the ModuleMetadata stating whether we have emitted the necessary refs?

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs (hunk 3 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs]: Yeah that's a good idea. I expect we'll always be loading with the full name, like below, so `args.Name` should contain the full name

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 4 above)
Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: I'm thinking about the potential case where we may fail to emit or store refs for some reason, should we keep trying for every single JIT?

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: That's a good point, we should be able to defer that until we are about to replace it! I'll add that to this PR.

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: I didn't refresh so I only saw your first comment when responding.

I like the idea of only attempting to emit the refs once, so I'll look into implementing that. Originally we stored the ref only once during `ModuleLoadFinished` and then just retrieved it in `JITCompilationStarted` so that wasn't an issue. We would just exit nicely if we couldn't find it.

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs (hunk 3 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs]: Added in 1e892a61569c1df2ec49af867ae9920e2c70aba7

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 4 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Implemented in 3aa43a974f95d25712fe8af5d728008a5409afbc

Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Looks great!

File: samples/Samples.HttpMessageHandler/Program.cs [hunk 5]
```diff
@@ -75,6 +75,9 @@ public static void Main(string[] args)
 
         private static async Task SendHttpClientRequestAsync(bool tracingDisabled)
         {
+            // Insert a call to the Tracer.Instance to include an AssemblyRef to Datadog.Trace assembly in the final executable
```
Comment (lucaspimentel) [inline samples/Samples.HttpMessageHandler/Program.cs]: Why do we need this in this sample app and not the others?

File: samples/Samples.SqlServer/Properties/launchSettings.json [hunk 6]
```diff
@@ -11,6 +11,7 @@
         "CORECLR_PROFILER": "{846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
         "CORECLR_PROFILER_PATH": "$(ProjectDir)$(OutputPath)profiler-lib\\Datadog.Trace.ClrProfiler.Native.dll",
 
+        "DD_DOTNET_TRACER_HOME": "$(ProjectDir)$(OutputPath)profiler-lib",
```
Comment (lucaspimentel) [inline samples/Samples.SqlServer/Properties/launchSettings.json]: Don't we need to add this in every sample app?

File: samples/Samples.SqlServer/Properties/launchSettings.json [hunk 7]
```diff
@@ -11,6 +11,7 @@
...
         "DD_INTEGRATIONS": "$(ProjectDir)$(OutputPath)profiler-lib\\integrations.json"
```
Comment (lucaspimentel) [inline samples/Samples.SqlServer/Properties/launchSettings.json]: Should we drop `DD_INTEGRATIONS` and figure it out from `DD_DOTNET_TRACER_HOME` instead? (Can be separate PR)

File: samples/Samples.HttpMessageHandler/Program.cs (hunk 5 above)
Comment (zacharycmontoya) [reply samples/Samples.HttpMessageHandler/Program.cs]: In this particular case I was trying to test a reference to just `Datadog.Trace` and not our `Datadog.Trace.ClrProfiler.Managed` assembly, to see if there would be any weird interactions with `Datadog.Trace.dll` already being in the bin folder. We do not need to have it there, but I thought it would be an interesting test.

File: samples/Samples.SqlServer/Properties/launchSettings.json (hunk 6 above)
Comment (zacharycmontoya) [reply samples/Samples.SqlServer/Properties/launchSettings.json]: Yes we'll need this for all apps so we can test in VS. This PR only converts a handful of the sample apps to the new mechanism so that we are still testing the "NuGet" scenario where the assemblies are loaded from the bin directory and do not invoke our AssemblyResolve event. Should I convert all sample apps in this PR to use this approach?

File: samples/Samples.SqlServer/Properties/launchSettings.json (hunk 7 above)
Comment (zacharycmontoya) [reply samples/Samples.SqlServer/Properties/launchSettings.json]: Is there anywhere where we might need the flexibility of an integrations file outside of the profiler directory? Perhaps we could default to the DD_DOTNET_TRACER_HOME directory and use DD_INTEGRATIONS as a sort of override?

File: samples/Samples.SqlServer/Properties/launchSettings.json (hunk 6 above)
Comment (lucaspimentel) [reply samples/Samples.SqlServer/Properties/launchSettings.json]: I just worried that when this PR is merged, running some apps from VS will load the profiler and others will not. Then we spend time troubleshooting the profiler until we realize it's a missing env var.

Comment (zacharycmontoya) [reply samples/Samples.SqlServer/Properties/launchSettings.json]: Ok, I can add the DD_DOTNET_TRACER_HOME env var to all of the `launchSettings.json` files so we don't run into that issue.

File: samples/Samples.SqlServer/Properties/launchSettings.json (hunk 7 above)
Comment (lucaspimentel) [reply samples/Samples.SqlServer/Properties/launchSettings.json]: > default to the `DD_DOTNET_TRACER_HOME` directory and use `DD_INTEGRATIONS` as a sort of override

I like this idea. The latter also lets us define multiple json files. We should probably also rename that to `DD_DOTNET_INTEGRATIONS` at some point.

Let's leave as-is for this PR and come back to that later.

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            var tracerHomeDirectory = Environment.GetEnvironmentVariable("DD_DOTNET_TRACER_HOME") ?? string.Empty;
+            return Path.Combine(tracerHomeDirectory, tracerFrameworkDirectory);
+        }
+
+        private static Assembly AssemblyResolve_ManagedProfilerDependencies(object sender, ResolveEventArgs args)
+        {
+            var assemblyName = new AssemblyName(args.Name);
+            var path = Path.Combine(ManagedProfilerDirectory, $"{assemblyName.Name}.dll");
+
+            if (assemblyName.Name.Contains("Datadog.Trace")
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs]: nit: `StartsWith()` can be slightly faster than `Contains()` (worst case, the latter has to look at the entire string).

In either case, we should probably specify `StringComparer.OrdinalIgnoreCase`/`StringComparison.OrdinalIgnoreCase`.

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs [hunk 9]
```diff
@@ -6,29 +6,31 @@ namespace Datadog.Trace.ClrProfiler.Managed.Loader
...
 
-        internal static bool ManagedAssemblyFound { get; set; }
+        internal static string ManagedProfilerDirectory { get; }
 
-        private static void LoadManagedAssembly()
+        private static bool TryLoadManagedAssembly()
         {
             try
             {
                 Assembly.Load(new AssemblyName("Datadog.Trace.ClrProfiler.Managed, Version=1.7.0.0, Culture=neutral, PublicKeyToken=def86d061d0d2eeb"));
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs]: Should we use the assembly version from `DatadogTracer.ClrProfiler.Managed.Loader` here instead of hard-coding it? (assuming they will always match)

```suggestion
                var assemblyName = Assembly.GetExecutingAssembly().GetName();
                assemblyName.Name = "Datadog.Trace.ClrProfiler.Managed";
                
                Assembly.Load(assemblyName);
```

File: samples/Samples.SqlServer/Properties/launchSettings.json (hunk 6 above)
Comment (zacharycmontoya) [reply samples/Samples.SqlServer/Properties/launchSettings.json]: Fixed in be14e00afdacd21136a502cc8bf57e8e1530db60. Also, this edits the MSI to add the `DD_DOTNET_TRACER_HOME` environment variable which I forgot to do before.

File: test/Datadog.Trace.TestHelpers/EnvironmentHelper.cs [hunk 10]
```diff
@@ -201,18 +201,25 @@ public static void ClearProfilerEnvironmentVariables()
         {
             var processName = processPath;
             string profilerEnabled = _requiresProfiling ? "1" : "0";
+            string profilerPath;
```
Comment (lucaspimentel) [inline test/Datadog.Trace.TestHelpers/EnvironmentHelper.cs]: nit: looks like you can assign this here and remove the ones below
```suggestion
            string profilerPath = GetProfilerPath();
```

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs (hunk 9 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs]: That's a good idea, I hadn't thought of that before. The version should always be incremented with the native (and managed) profiler so I'm happy to make this change.

File: test/Datadog.Trace.TestHelpers/EnvironmentHelper.cs (hunk 10 above)
Comment (zacharycmontoya) [reply test/Datadog.Trace.TestHelpers/EnvironmentHelper.cs]: Good catch, I'll fix that.

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs (hunk 9 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs]: Fixed in 7a2a2ce4d640c750eca578a28ef7c32cde1eadb4 .

File: test/Datadog.Trace.TestHelpers/EnvironmentHelper.cs (hunk 10 above)
Comment (zacharycmontoya) [reply test/Datadog.Trace.TestHelpers/EnvironmentHelper.cs]: Fixed in 7a2a2ce4d640c750eca578a28ef7c32cde1eadb4 .

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs (hunk 8 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs]: Fixed in 7a2a2ce4d640c750eca578a28ef7c32cde1eadb4 .

File: src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+            return Path.Combine(tracerHomeDirectory, tracerFrameworkDirectory);
+        }
+
+        private static Assembly AssemblyResolve_ManagedProfilerDependencies(object sender, ResolveEventArgs args)
+        {
+            var assemblyName = new AssemblyName(args.Name);
+            var path = Path.Combine(ManagedProfilerDirectory, $"{assemblyName.Name}.dll");
+
+            if (assemblyName.Name.StartsWith("Datadog.Trace", StringComparison.OrdinalIgnoreCase)
+                && assemblyName.FullName.Contains("PublicKeyToken=def86d061d0d2eeb")
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs]:
```suggestion
                && assemblyName.FullName.Contains("PublicKeyToken=def86d061d0d2eeb", StringComparer.OrdinalIgnoreCase)
```

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs]: Fixed in b9c80c289bb5e108f1c25d9973cb6f669659f975 . It turns out `string.Contains` doesn't have a built-in method that takes a `string` and a `StringComparer` so I had to resort to calling `string.IndexOf` which does take those arguments

Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs]: Oops, my bad! I was getting Intellisense for `IEnumerable<char>.Contains(...)` 🤦‍♂
